File: 875. Koko Eating Bananas.cpp
#include <vector>
#include <climits>
using namespace std;
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int max = INT_MIN;
        for(int i = 0; i < piles.size(); i++){
            if(max < piles[i]) max = piles[i];
        }
        int left = 1, right = max;
        int ans = INT_MAX;
        while(left < right){
            int mid = left + (right - left)/2;
            int calc = 0;
            for (int pile : piles) {
                calc += (pile + mid - 1) / mid;
            }
            if(calc <= h){
                ans = mid;
                right = mid;
            }
            else if(calc > h){
                left = mid + 1;
            }
        }
        if(ans == INT_MAX) return left;
        return ans;
    }
};

File: 875. Koko Eating Bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int n = piles.length;
        int maxPile = piles[0];

        for(int pile : piles) if(maxPile < pile) maxPile = pile;
        
        if(h == n) return maxPile;
        int left = 1, right = maxPile, minSpeed = maxPile;
        while(left < right){
            int mid = left + (right - left)/2;

            int time = 0;
            for(int pile : piles){
                time += (pile + mid - 1) / mid;
            }
            if(time > h) left = mid + 1;
            else{
                minSpeed = mid;
                right = mid;
            }
        }
        return minSpeed;


    }
}